managedservices/managedservices.go

```diff
@@ -0,0 +1,246 @@
+	ds, err := ms.hostFramework.K8sClient().Apps().DaemonSets(expectedDaemonSet.Namespace).Get(expectedDaemonSet.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return microerror.Maskf(notFoundError, "daemonset %#q", expectedDaemonSet.Name)
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = ms.checkLabels("daemonset labels", expectedDaemonSet.Labels, ds.ObjectMeta.Labels)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = ms.checkLabels("daemonset matchLabels", expectedDaemonSet.MatchLabels, ds.Spec.Selector.MatchLabels)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = ms.checkLabels("daemonset pod labels", expectedDaemonSet.Labels, ds.Spec.Template.ObjectMeta.Labels)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// checkDeployment ensures that key properties of the deployment are correct.
+func (ms *ManagedServices) checkDeployment(expectedDeployment Deployment) error {
+	ds, err := ms.hostFramework.K8sClient().Apps().Deployments(expectedDeployment.Namespace).Get(expectedDeployment.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return microerror.Maskf(notFoundError, "deployment: %#q", expectedDeployment.Name)
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if int32(expectedDeployment.Replicas) != *ds.Spec.Replicas {
+		return microerror.Maskf(invalidReplicasError, "expected %d replicas got: %d", expectedDeployment.Replicas, *ds.Spec.Replicas)
+	}
+
+	err = ms.checkLabels("deployment labels", expectedDeployment.Labels, ds.ObjectMeta.Labels)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = ms.checkLabels("deployment matchLabels", expectedDeployment.MatchLabels, ds.Spec.Selector.MatchLabels)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = ms.checkLabels("deployment pod labels", expectedDeployment.Labels, ds.Spec.Template.ObjectMeta.Labels)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (ms *ManagedServices) checkLabels(labelType string, expectedLabels, labels map[string]string) error {
+	if !reflect.DeepEqual(expectedLabels, labels) {
+		ms.logger.Log("level", "debug", "message", fmt.Sprintf("expected %s: %v got: %v", labelType, expectedLabels, labels))
+		return microerror.Maskf(invalidLabelsError, "%s do not match expected labels", labelType)
+	}
+
+	return nil
+}
+
+func validateChartConfig(chartConfig ChartConfig) error {
+	if chartConfig.ChannelName == "" {
+		return microerror.Maskf(invalidConfigError, "%T.ChannelName must not be empty", chartConfig)
+	}
+	if chartConfig.ChartName == "" {
+		return microerror.Maskf(invalidConfigError, "%T.ChartName must not be empty", chartConfig)
+	}
+	if chartConfig.Namespace == "" {
+		return microerror.Maskf(invalidConfigError, "%T.Namespace must not be empty", chartConfig)
+	}
+
+	return nil
+}
+
+func validateChartResources(chartResources ChartResources) error {
```

Comment:
I think the validation functions could actually just be methods on the types. Then we have the clean interface `Validate() error` which we also want to pursue for general config structures.

Reply:
Thanks, much nicer! Done.